### DISCUSSION TITLE
Why isn't the `searchParams` prop on Page components a `URLSearchParams` instance?

### DISCUSSION THREAD

**Discussion by jviall** [vercel/next.js#52013] [App Router]
Per the documentation here: https://nextjs.org/docs/app/api-reference/file-conventions/page#searchparams-optional
> `searchParams` returns a plain JavaScript object and not a `URLSearchParams` instance.

This seems like a poor API choice. The URLSearchParams API is well known, easy to use, and easy to reckon with in TypeScript projects. There's no documentation on this prop. Does Next decode the search params? If so, how? In this world, you have to document a proprietary API around search params, and we have to manually reckon with a super open-ended object type when using TS, which is frustrating and unnecessary.

Why not let us interact with the native API? Even if it's a limitation of the server, it seems like something that Next could implement for back-end environments to unify the experience of working with search params without much more effort.

**Comment by K1NXZ** [comment]
Yes. I also struggled with that. Its very inconsistent with the client hook `useSearchParams` which makes it impossible to reuse code.

**Reply by MicahParks** [reply]
If anyone knows of a good way to feed the `searchParams` type `{ [key: string]: string | string[] | undefined }` into a `new URLSearchParams(searchParams)`, please do leave a comment. I'm curious of the best way.

**Reply by jviall** [reply]
[Per the docs](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/URLSearchParams#options), it's trivial enough for the `URLSearchParams` constructor to parse a flat Record of string keys and string values:
```js
const paramsObj = {
  a: 'one',
  b: 'two',
};
const params = new URLSearchParams(paramsObj);
params.toString(); // 'a=one&b=two'
```

You can also pass an array of entries, again given that original object is a flat record of string keys and values:

```js
const params = new URLSearchParams([...Object.entries(paramsObj), ['c', 'three']]);
params.toString(); // 'a=one&b=two&c=three'
```

The difficulty comes with handling the `[key: string]: string[]` case, where if a url contains two entries of the same param (e.g. `a=first&a=second` which will be represented in `searchParams` as `{ a: ['first', 'second'] }`) it won't be correctly parsed by the `URLSearchParams` constructor in either of the above approaches because it is not a flat object of only string values.

```js
const params = new URLSearchParams(searchParams); // URL params are 'a=first&a=second'
params.toString(); // incorrect result of 'a=first%2Csecond'
```

This is the case that I believe Next should handle for us by providing `searchParams` as a `URLSearchParams` instance.
